docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,844 @@
+the old contract.
+
+Please note that we are planning to embed this sequence (pause - upgrade -
+unpause) as a part of `clusterctl upgrade apply` command when there is a
+contract change.
+
+
+##### Changing a provider
+
+On top of changing a provider version (upgrades), the operator supports also
+changing other provider fields, most notably controller flags and variables.
+
+The operation internally works like upgrades: The current instance of the
+provider is deleted, while preserving CRDs, namespaced and user objects A new
+instance of the provider is installed with the new set of flags/variables.
+
+Please note that clusterctl currently does not support this operation.
+
+See Example 1 in [Example API Usage](#example-api-usage)
+
+##### Deleting a provider
+
+In order to delete a provider you have to delete the corresponding provider
+object.
+
+Deletion of the provider will be blocked if any workload cluster using the
+provider still exists.
+
+Additionally, deletion of a core provider should be blocked if there are still
+other providers in the management cluster.
+
+#### Operator Lifecycle Management
+
+##### Operator Installation
+
+- `clusterctl init` will install the operator and its corresponding CRDs as a
+  pre-requisite if the operator doesn’t already exist. Please note that this
+  command will consider image overrides defined in the local clusterctl config
+  file.
+- If the admin does not want to use clusterctl to install the operator, it is
+  possible to kubectl apply the operator yaml that will be published in the
+  cluster-api release artifacts.
+- There can be only one operator installed per management cluster, across all
+  namespaces.
+
+##### Operator Upgrade
+
+- The admin can use `clusterctl upgrade operator` to upgrade the operator
+  components. Please note that this command will consider image overrides
+  defined in the local clusterctl config file.
+- If the admin doesn’t want to use clusterctl, they can use kubectl apply with
+  the latest version of the operator yaml that will be published in the
+  cluster-api release artifacts.
+- clusterctl will require a matching operator version. In the future, when
+  clusterctl move to beta/GA, we will reconsider supporting version skew
+  between clusterctl and the operator.
+
+##### Operator Delete
+
+- clusterctl will delete the operator as part of the `clusterctl delete --all`
+  command.
+- If the admin doesn’t want to use clusterctl, they can use kubectl delete.
+  However, it’s the admin’s responsibility to verify that there are no
+  providers running in the management cluster.
+
+#### Air gapped environment
+
+In order to use the operator in an air-gapped environment it is required
+
+1. Make the operator work in air-gapped environment
+   - To provide image overrides for the operator itself.
+   - TBD if operator yaml will be embedded in clusterctl or if it should be a
+     special artifact within the core provider repository.
```

Comment:
I would say the operator should have its own image, so I'm +1 to remove this TBD

Reply:
@fabriziopandini So we are leaning towards the operator having its own image and its own yaml that will be specified as part of the cluster-api release artifacts. Correct?

I wanted to confirm because I wanted to add a note about "how" we check that the operator needs an upgrade. 
That is, does the user get a newer version of clusterctl which has a newer version of the operator yaml embedded within it.
OR
For example, when the user does `clusterctl upgrade plan`, will clusterctl by default check the Cluster API releases and see if there is a newer version of CAPI and hence a newer version of the operator yaml.
In this scenario for the air-gapped environment, clusterctl can see the providers list in `~/.cluster-api/clusterctl.yaml` and use whatever repo is specified there.

_or we check both_ 🤷  WDYT?

Reply:
I'm assuming that the operator should be a separated package because we want to make it possible to use CAPI without the operator and/or without clusterctl, but might be worth to get other opinions here @vincepri @ncdc